### PR TITLE
Add Java parser

### PR DESCRIPTION
**Pull request by tonyyang-svail**
Part of #1126 

Next step:
1. Enable tests in our CI
2. Wrap Java Class as a command line

### CODE REVIEW

File: java/parser/README.md [hunk 1]
```diff
@@ -1,14 +1,13 @@
-# Build and Run Remote Parsers
+# Build and Test Third Party Parsers
 
-Currently, we have two remote parsers, each as a gRPC server in Java:
+We build and test two third party parsers:
 
 - hiveql
```
Comment (wangkuiyi) [inline java/parser/README.md]: HiveQL

File: java/parser/README.md [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
 - calcite
```
Comment (wangkuiyi) [inline java/parser/README.md]: Calcite

File: java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    parse_result.Statements = new ArrayList<String>();
+    parse_result.Position = -1;
+    parse_result.Error = "";
+
+    int accumulated_position = 0;
+    while (true) {
+      try {
+        SqlParser parser = SqlParser.create(sql);
+        SqlNode sqlnode = parser.parseQuery();
+        logger.debug("1 ----------------");
```
Comment (wangkuiyi) [inline java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java]: Remove debug logs.

File: java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        try {
+          SqlParser parser = SqlParser.create(sql.substring(0, epos));
+          SqlNode sqlnode = parser.parseQuery();
+
+          // parseQuery doesn't throw exception
+          parse_result.Statements.add(sql.substring(0, epos));
+
+          // multiple SQL statements
+          if (sql.charAt(epos) == ';') {
+            logger.debug("2.1 -----------------");
```
Comment (wangkuiyi) [inline java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java]: Remove debug logs.

File: java/parser/src/main/java/org/sqlflow/parser/ParseResult.java [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package org.sqlflow.parser;
+
+import java.util.ArrayList;
+
+public class ParseResult {
```
Comment (wangkuiyi) [inline java/parser/src/main/java/org/sqlflow/parser/ParseResult.java]: We need a comment explaining the use of this class.

File: java/parser/src/test/java/org/sqlflow/parser/ParserTest.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
+package org.sqlflow.parser;
+
+import static org.junit.Assert.*;
+
+import java.util.ArrayList;
+import org.junit.Test;
+
+public class ParserTest {
+  @Test
+  public void testCalciteParseAndSplit() {
```
Comment (wangkuiyi) [inline java/parser/src/test/java/org/sqlflow/parser/ParserTest.java]: It looks to me that the test for Calcite parser and the one for HiveQL should be in different test source files.

File: java/parser/README.md (hunk 1 above)
Comment (tonyyang-svail) [reply java/parser/README.md]: Done.

File: java/parser/README.md (hunk 2 above)
Comment (tonyyang-svail) [reply java/parser/README.md]: Done.

File: java/parser/src/main/java/org/sqlflow/parser/ParseResult.java (hunk 5 above)
Comment (tonyyang-svail) [reply java/parser/src/main/java/org/sqlflow/parser/ParseResult.java]: I've added the comments.

File: java/parser/src/test/java/org/sqlflow/parser/ParserTest.java (hunk 6 above)
Comment (tonyyang-svail) [reply java/parser/src/test/java/org/sqlflow/parser/ParserTest.java]: You are right. I have put two tests into two source files.

File: java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java (hunk 4 above)
Comment (tonyyang-svail) [reply java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java]: Done.

File: java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java (hunk 3 above)
Comment (tonyyang-svail) [reply java/parser/src/main/java/org/sqlflow/parser/CalciteParserAdaptor.java]: Done.

File: java/parser/src/test/java/org/sqlflow/parser/HiveQLParserAdaptorTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
+package org.sqlflow.parser;
+
+import static org.junit.Assert.*;
+
+import java.util.ArrayList;
+import org.junit.Test;
+
+public class HiveQLParserAdaptorTest {
```
Comment (typhoonzero) [inline java/parser/src/test/java/org/sqlflow/parser/HiveQLParserAdaptorTest.java:8]: These two tests have some code in common, maybe construct a base class to store the test case, and result testing functions and the sub-classes call the corresponding parser.

File: java/parser/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# Note: I used the official maven Docker image to build this package.
+# However, Maven complains it cannot find tools.jar. I verified that
+# I'd have to yum install java-8-openjdk in that Docker image to make
+# Maven works.  However, I'd follow this alternative to create a
+# Docker image from Ubuntu instead of Oracle Linux.
+FROM ubuntu:18.04
+
+RUN apt-get update && apt-get install -y openjdk-8-jdk maven openjfx
+ENV JAVA_HOME=/usr/lib/jvm/java-8-openjdk-amd64
```
Comment (typhoonzero) [inline java/parser/Dockerfile:9]: Are you going to delete this Dockerfile once the parser development is done? We can use the SQLFlow Docker image to run these tests.

**Review by typhoonzero — APPROVED** [review]
LGTM basically.

File: java/parser/Dockerfile (hunk 8 above)
Comment (tonyyang-svail) [reply java/parser/Dockerfile:9]: Yep. I will do so in the next PR. :)

File: java/parser/src/test/java/org/sqlflow/parser/HiveQLParserAdaptorTest.java (hunk 7 above)
Comment (tonyyang-svail) [reply java/parser/src/test/java/org/sqlflow/parser/HiveQLParserAdaptorTest.java:8]: Sure. I will refactorize it in the next PR.

**Pull request merged**